File: leetcode/156.binary-tree-upside-down.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):

    def __init__(self, *args, **kwargs):
        self.newRoot = None

    def invert(self, node):
        if node is None:
            return

        if node.left is None:
            self.newRoot = node
        else:
            self.invert(node.left)

            node.left.left = node.right
            node.left.right = node
            node.left = None
            node.right = None
        
    def upsideDownBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        self.invert(root)
        return self.newRoot

File: leetcode/156.binary-tree-upside-down.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {

private:
    TreeNode *newRoot = NULL;

    void invert(TreeNode *cur)
    {
        if (cur == NULL)
        {
            return;
        }

        if (cur->left == NULL)
        {
            newRoot = cur;
        }
        else
        {
            invert(cur->left);
            cur->left->left = cur->right;
            cur->left->right = cur;
            cur->left = NULL;
            cur->right = NULL;
        }
    }
public:
    TreeNode* upsideDownBinaryTree(TreeNode* root) {
        invert(root);
        return newRoot;
    }
};
